File: elixir/pe023.exs
defmodule ProjectEuler do
  def pe023(n) do
    {list, map} = compute_data(n, [], %{})
    map[hd(list)]
    start_sum = start_vals(hd(list) - 1, 0)
    check_each_value(n, hd(list), list, map, start_sum)
  end
   
  defp check_each_value(n, value, list, map, sum) when value < n do
    new_sum = check_each_abundant_value(value, list, map, sum)
    check_each_value(n, value + 1, list, map, new_sum)
  end

  defp check_each_value(_, _, _, _, sum) do
    sum
  end

  defp check_each_abundant_value(value, [head|tail], map, sum) do
    if head >= value do
      check_each_abundant_value(value, [], map, sum + value)
    else
      if Map.has_key?(map, value - head) do
        check_each_abundant_value(value, [], map, sum)
      else
        check_each_abundant_value(value, tail, map, sum)
      end
    end
  end

  defp check_each_abundant_value(_, [], _, sum) do
    sum
  end

  defp start_vals(x, sum) when x > 0 do
    start_vals(x - 1, sum + x)
  end
   
  defp start_vals(_, sum) do
    sum
  end
   
  defp compute_data(n, list, map) when n > 0 do
    root = Kernel.round :math.sqrt(n) + 1
    fsum = sum_factors(n, root, 2, 1)
    if n < fsum do
      adict = Map.put map, n, fsum
      compute_data(n - 1, [n | list], adict)
    else
      compute_data(n - 1, list, map)
    end
  end

  defp compute_data(_, list, map) do
    {list, map}
  end

  defp sum_factors(n, root, count, sum) when count < root do
    if rem(n, count) == 0 do
      divisor = div(n, count)
      if count == divisor do
        sum_factors(n, root, count + 1, sum + count)
      else
        sum_factors(n, root, count + 1, sum + count + divisor)
      end
    else
      sum_factors(n, root, count + 1, sum)
    end
  end

  defp sum_factors(_, _, _, sum) do
    sum
  end
end

IO.puts(ProjectEuler.pe023(28123))

File: haskell/pe023.hs
-- this computes projecteuler.net problem 023

import qualified Data.Map as Map

check_each_abundant_value :: Int -> [Int] -> Map.Map Int Int -> Int -> Int
check_each_abundant_value value (x:xs) mmap sum =
  do
    if x > value then
      check_each_abundant_value value [] mmap (sum + value)
    else
      if Map.member (value - x) mmap then
        check_each_abundant_value value [] mmap sum 
      else
        check_each_abundant_value value xs mmap sum 

check_each_abundant_value _ [] _ sum = sum

start_vals :: Int -> Int -> Int
start_vals x sum | x > 0 = start_vals (x - 1) (sum + x)
start_vals x sum = sum

check_each_value :: Int -> Int -> [Int] -> Map.Map Int Int -> Int -> Int
check_each_value n value list mmap sum | value < n = 
  do
    let new_sum = check_each_abundant_value value list mmap sum
    check_each_value n (value + 1) list mmap new_sum

check_each_value _ _ _ _ sum = sum

sum_factors :: Int -> Int -> Int -> Int -> Int
sum_factors n root count sum | count < root =
  do
    if rem n count == 0 then
      do
        let divisor = div n count
        if count == divisor then
          sum_factors n root (count + 1) (sum + count)
        else
          sum_factors n root (count + 1) (sum + count + divisor)
    else
      sum_factors n root (count + 1) sum

sum_factors _ _ _ sum = sum

compute_data :: Int -> Map.Map Int Int -> Map.Map Int Int
compute_data n mmap | n > 0 = 
  do
    let root = ceiling (sqrt ((fromIntegral n) + 1))
    let fsum = sum_factors n root 2 1

    if n < fsum then
      compute_data (n - 1) (Map.insert n fsum mmap)
    else
      compute_data (n - 1) mmap

compute_data _ mmap = mmap

pe023 :: Int -> Int
pe023 n = do
  let mmap = compute_data n Map.empty
  let keys = Map.keys mmap
  let first = head keys
  let start_sum = start_vals (first - 1) 0
  check_each_value n first keys mmap start_sum

main = print(pe023 28123)
